noir-projects/aztec-nr/aztec/src/context/interface.nr

```diff
@@ -55,3 +61,151 @@ trait PublicContextInterface {
+        context.call_private_function_with_packed_args(self.target_contract, self.selector, self.args_hash, true, false).assert_empty();
+    }
+
+    pub fn delegate_call(self, context: &mut PrivateContext) {
+        context.call_private_function_with_packed_args(self.target_contract, self.selector, self.args_hash, false, true).assert_empty();
+    }
+}
+
+struct PublicCallInterface<T> {
+    target_contract: AztecAddress,
+    selector: FunctionSelector,
+    args_hash: Field,
+}
+
+impl<T> PublicCallInterface<T> {
+
+    pub fn call<N>(self, context: &mut PublicContext) -> T where T: Deserialize<N> {
+        let returns = context.call_public_function_with_packed_args(
+            self.target_contract,
+            self.selector,
+            self.args_hash,
+            false,
+            false
+        );
+        returns.deserialize_into()
+    }
+
+    pub fn static_call<N>(self, context: &mut PublicContext) -> T where T: Deserialize<N> {
+        let returns = context.call_public_function_with_packed_args(self.target_contract, self.selector, self.args_hash, true, false);
+        returns.deserialize_into()
+    }
+
+    pub fn delegate_call<N>(self, context: &mut PublicContext) -> T where T: Deserialize<N> {
+        let returns = context.call_public_function_with_packed_args(self.target_contract, self.selector, self.args_hash, false, true);
+        returns.deserialize_into()
+    }
+
+    pub fn enqueue(self, context: &mut PrivateContext) {
+        context.call_public_function_with_packed_args(
+            self.target_contract,
+            self.selector,
+            self.args_hash,
+            false,
+            false
+        )
+    }
+
+    pub fn static_enqueue(self, context: &mut PrivateContext) {
+        context.call_public_function_with_packed_args(self.target_contract, self.selector, self.args_hash, true, false)
+    }
+
+    pub fn delegate_enqueue(self, context: &mut PrivateContext) {
+        context.call_public_function_with_packed_args(self.target_contract, self.selector, self.args_hash, false, true)
+    }
+}
+
+struct PublicVoidCallInterface {
+    target_contract: AztecAddress,
+    selector: FunctionSelector,
+    args_hash: Field
+}
+
+impl PublicVoidCallInterface {
+    pub fn call(self, context: &mut PublicContext) {
+        context.call_public_function_with_packed_args(
+            self.target_contract,
+            self.selector,
+            self.args_hash,
+            false,
+            false
+        ).assert_empty()
+    }
+
+    pub fn static_call(self, context: &mut PublicContext) {
+        context.call_public_function_with_packed_args(self.target_contract, self.selector, self.args_hash, true, false).assert_empty();
+    }
+
+    pub fn delegate_call(self, context: &mut PublicContext) {
+        context.call_public_function_with_packed_args(self.target_contract, self.selector, self.args_hash, false, true).assert_empty();
+    }
+
+    pub fn enqueue(self, context: &mut PrivateContext) {
```

Comment:
I think it's nice to have this distinction now. The compiler will also prevent devs from doing the wrong thing based on parameters.